book/src/rfc/const_generic.md

```diff
@@ -0,0 +1,265 @@
+
+For example:
+```rust
+houses[1].rooms[2].size
+```
+
+The variable `houses` holds all the `cvars` to represent the underlying low level variables. If `houses` is a mutable variable, it will be a reference, which records the following instead of holding the actual `cvars` array:
+ - the variable name `houses` in local scope
+ - start index of the `cvars` to represent the current access path
+ - the length of the `cvars` to represent size of the element targeted by the current access path
+
+So when it wants to update the elements inside the `houses`, it will need to reference the start index and the length of the `cvars` of the elements, so that it can narrow down the scope of the `cvars` to be updated.
+
+The following code is how it narrows down the scope for a field access or array access:
+
+```rust
+pub(crate) fn narrow(&self, start: usize, len: usize) -> Self {
+    match self {
+        VarOrRef::Var(var) => {
+            let cvars = var.range(start, len).to_vec();
+            VarOrRef::Var(Var::new(cvars, var.span))
+        }
+
+        //      old_start
+        //      |
+        //      v
+        // |----[-----------]-----| <-- var.cvars
+        //       <--------->
+        //         old_len
+        //
+        //
+        //          start
+        //          |
+        //          v
+        //      |---[-----]-|
+        //           <--->
+        //            len
+        //
+        VarOrRef::Ref {
+            var_name,
+            start: old_start,
+            len: old_len,
+        } => {
+            // ensure that the new range is contained in the older range
+            assert!(start < *old_len); // lower bound
+            assert!(start + len <= *old_len); // upper bound
+            assert!(len > 0); // empty range not allowed
+
+            Self::Ref {
+                var_name: var_name.clone(),
+                start: old_start + start,
+                len,
+            }
+        }
+    }
+}
+```
+
+This allows to find the range of the `cvars` belonging to the element targeted by the current access path under the local variable behind the `var_name`, which in this case is the `houses`. 
+
+But it doesn't keep track of the structure of the data. For example, when we are accessing `houses[1].rooms[2]`, it should be able to check if the access is within the array bounds. Although the current design can check if it is out of bound by checking access index against the total number of `cvars` for `houses`, it can't label which element access is out of bound as it could be either out of bound of `houses[1]` or the nested access `rooms[2]`.
+
+### Introduce ComputedExpr
+
+With the const variable as array size, we need to propagate its value during circuit generation to check if the access is within the bounds of the array. So we need a way to retain structure information of a variable that is currently represented by `VarOrRef`.
```

Comment:
not sure I understand why, if it's not in the bounds of the array you can see that pretty quickly as `some_array.get(some_idx)` will return `None`

Reply:
I probably need to rephrase this one also. But I meant during `ExprKind::Array` it was using `size_of` for a type to determine the size of the array. With the support of const variable length for array, the actual value of the const variable needs to be propagated to `ExprKind::Array` computation.

Reply:
It is because it can't rely on the `size_of` to do the check anymore for arrays with variable length. It was using `size_of` to determine the size of `some_array`.

I detailed the explain in the `Resolving const variables`.